pkg/services/hcloud/server/server.go

```diff
@@ -92,42 +105,97 @@ func (s *Service) Reconcile(ctx context.Context) (res reconcile.Result, err erro
 
 	conditions.MarkTrue(s.scope.HCloudMachine, infrav1.BootstrapReadyCondition)
 
-	// try to find an existing server
-	server, err := s.findServer(ctx)
-	if err != nil {
-		return reconcile.Result{}, fmt.Errorf("failed to get server: %w", err)
+	var server *hcloud.Server
+	if s.scope.HCloudMachine.Spec.ProviderID != nil {
+		server, err = s.findServer(ctx)
+		if err != nil {
+			return reconcile.Result{}, fmt.Errorf("findServer: %w", err)
+		}
+
+		if server == nil {
+			// The server did disappear in HCloud? Maybe it was delete via web-UI.
+			// We set MachineError. CAPI will delete machine.
+			msg := fmt.Sprintf("hcloud server (%q) no longer available. Setting MachineError.",
+				*s.scope.HCloudMachine.Spec.ProviderID)
+
+			ctrl.LoggerFrom(ctx).Error(errors.New(msg), msg,
+				"ProviderID", *s.scope.HCloudMachine.Spec.ProviderID,
+				"BootState", s.scope.HCloudMachine.Status.BootState,
+				"BootStateSince", s.scope.HCloudMachine.Status.BootStateSince,
+			)
+
+			s.scope.SetError(msg, capierrors.CreateMachineError)
+			s.scope.HCloudMachine.SetBootState(infrav1.HCloudBootStateUnset)
+			record.Warn(s.scope.HCloudMachine, "NoHCloudServerFound", msg)
+
+			// no need to requeue.
+			return reconcile.Result{}, nil
+		}
+		updateHCloudMachineStatusFromServer(s.scope.HCloudMachine, server)
```

Comment:
why do you wanna update the status here already? Shouldn't we do that in the main  handle functions?